### DISCUSSION TITLE
DDD e EF Core

### DISCUSSION THREAD

**Discussion by MateusMMontagnoli** [balta-io/forum#697] [Tira Dúvidas] [answered]
Boa noite, tudo bem?

Estou me aventurando na parte da modelação das minhas entidades e acabei me esbarrando em um problema com o EF Core. Boa parte da minha base de dados não trabalha com UUIDs ou Id incrementais, por exemplo: Digamos que tenho uma tabela para PessoaJuridica, elá não terá um PessoaId ou PessoaGuid, o que a identifica é o Cnpj. Outras tabelas, seguem o mesmo principio, e até possuindo PKs compostas. 

Normalmente nos exemplos de entidades é sempre mostrado com elas com uma UUId ou um Id incremental, como trabalharia com entidades no formato do banco falado? Tentei utilizar classes como Ids para tentar abstrair mas estou falhando na hora de configurar com o EF Core, onde o entity não está conseguindo transcrever o modelo do banco para entidade. 

Abaixo tem parte da implementação que fiz:

```C#
public abstract class Entity<TId>
{
    public TId Id { get; protected set; }

   protected Entity()
   {
       Id = default!;
   }

    protected Entity(TId id)
    {
        Id = id;
    }
}

```

```C#
public class Checklist : Entity<ChecklistId>
{
    public Checklist(ChecklistId id, string observation, int userId, DateTime createdAt)
    {
        Id = id;
        CreatedAt = createdAt;
        Observation = observation;
        UserId = userId;
    }

    public ChecklistId Id { get; set; }
    public string Observation { get; set; }
    public int UserId { get; set; }
    public DateTime CreatedAt { get; set; }
}
```

Classe do Id

```C#
public class ChecklistId
{
    public ChecklistId(int inspectionId, string questionGroup, string question, string answer)
    {
        InspectionId = inspectionId;
        QuestionGroup = questionGroup;
        Question = question;
        Answer = answer;
    }

    public int InspectionId { get; set; }
    public string QuestionGroup { get; set; }
    public string Question { get; set; }
    public string Answer { get; set; }
}
```

Configuração do EF Core

```C#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    base.OnModelCreating(modelBuilder);

    modelBuilder.Entity<Checklist>(entity =>
    {
        entity.HasKey(e => new { e.Id.InspectionId, e,Id.QuestionGroup, e.Id.Question, e.Id.Answer });

        entity.OwnsOne(e => e.Id, id =>
        {
            id.Property(i => i.InspectionId).HasColumnName("InspectionId");
            id.Property(i => i.QuestionGroup).HasColumnName("QuestionGroup");
            id.Property(i => i.Question).HasColumnName("Question");
            id.Property(i => i.Answer).HasColumnName("Answer");
        });

        // ... outras propriedades
    });
}
```

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @MateusMMontagnoli, como vai?

Neste cenário, ou você usa o ID como String e informa o CPF/CNPJ ali, ou cria uma especialização para elas.
Acredito que a primeira opção é a mais simples!

Ou melhor ainda, cria um ID nestas tabelas... nada impede você buscar e identificar pelo documento também...

**Reply by MateusMMontagnoli** [reply]
Opa, tranquilo e você? @andrebaltieri 
Acredito que entendi.
Por exemplo, na questão da Entidade Checklist, você acha que seria interessante criar um ID para a tabela e criar uma Unique Constraint para (inspectionId, questionGroup, question,answer)? o mesmo digamos para as outras tabelas que possuo com várias chaves compostas?
Acredito que a criação de Ids na tabelas ajudaria muito.
Mas, o meu medo é que ao criar Ids eu afetaria a normalização do banco.

**Reply by andrebaltieri** [reply]
Isso ai!
